server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+/*
+ * Tencent is pleased to support the open source community by making
+ * Firestorm-Spark remote shuffle server available. 
+ *
+ * Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved. 
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.tencent.rss.server.buffer;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.tencent.rss.common.BufferSegment;
+import com.tencent.rss.common.ShuffleDataResult;
+import com.tencent.rss.common.ShufflePartitionedBlock;
+import com.tencent.rss.common.ShufflePartitionedData;
+import com.tencent.rss.common.util.Constants;
+import com.tencent.rss.server.ShuffleDataFlushEvent;
+import com.tencent.rss.server.ShuffleFlushManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+public class ShuffleBuffer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ShuffleBuffer.class);
+
+  private final long capacity;
+  private long size;
+  // blocks will be added to inFlushBlockMap as <eventId, blocks> pair
+  // it will be removed after flush to storage
+  // the strategy ensure that shuffle is in memory or storage
+  private List<ShufflePartitionedBlock> blocks;
+  private Map<Long, List<ShufflePartitionedBlock>> inFlushBlockMap;
+  private static final long INVALID_EVENT_ID = -1;
+
+  public ShuffleBuffer(long capacity) {
+    this.capacity = capacity;
+    this.size = 0;
+    this.blocks = new LinkedList<>();
+    this.inFlushBlockMap = Maps.newConcurrentMap();
+  }
+
+  public int append(ShufflePartitionedData data) {
+    int mSize = 0;
+
+    synchronized (this) {
+      for (ShufflePartitionedBlock block : data.getBlockList()) {
+        blocks.add(block);
+        mSize += block.getSize();
+        size += mSize;
+      }
+    }
+
+    return mSize;
+  }
+
+  public synchronized ShuffleDataFlushEvent toFlushEvent(
+      String appId,
+      int shuffleId,
+      int startPartition,
+      int endPartition,
+      Map<String, Map<Integer, AtomicLong>> shuffleSizeMap,
```

Comment:
I think we should only pass a size not a map.

Reply:
The map is used to track shuffleId -> dataSize, it should be updated when the size is changed.
ShuffleBuffer has to get this map, and pass it as a parameter is one of ways.

Reply:
One Buffer only care the shuffle size of itself.

Reply:
ok, size update can be called after `buffer.toFlushEvent` in `ShuffleBufferManager`